dbms/src/Flash/Coprocessor/JoinInterpreterHelper.h

```diff
@@ -36,8 +37,50 @@ struct JoinKeyType
 };
 using JoinKeyTypes = std::vector<JoinKeyType>;
 
-struct JoinOtherConditions
+/// JoinNonEqualConditions is the join conditions that are not evaluated in the "hash table probe" stage, there are two kinds of these conditions
+/// 1. conditions that need to be evaluated before "hash table probe" stage
+///    a. left conditions: the filter condition on the left table
+///    b. right conditions: the filter condition on the right table
+/// 2. conditions that need to be evaluated after "hash table probe" stage
+///    a. other condition: the filter that involve columns from both left and right table
+///    b. other condition from in: a special case that convert from `not in subquery`
+///    c. null aware equal condition: used for null aware join
+/// some examples
+/// Left condition:
+///  `select * from t1 left join t2 on t1.col1 = t2.col2 and t1.col2 > 1`
+///   - left condition is `t1.col2 > 1`
+///   note left condition should only exists in left/semi/left-semi join, otherwise, `t1.col2 > 1` should be pushed down as a filter on table scan
+/// Right condition:
+///  `select * from t1 right join t2 on t1.col1 = t2.col2 and t2.col2 > 1`
+///   - right condition is `t2.col2 > 1`
+///   note left condition should only exists in right join, otherwise, `t2.col2 > 1` should be pushed down as a filter on table scan
+///
+/// Other condition, other condition from in and null aware equal condition:
+/// Example 1:
+///   `select * from t1 inner join t2 on t1.col1 = t2.col1 and t1.col2 > t2.col2`
+///   - join key are `t1.col1` and `t2.col1`
+///   - other_cond is `t1.col2 > t2.col2`
+///
+/// Example 2:
+///   `select * from t1 where col1 not in (select col1 from t2 where t1.col2 = t2.col2 and t1.col3 > t2.col3)`
+///   There are several possibilities which depends on TiDB's planner.
+///   1. cartesian anti semi join.
+///      - join key is empty
+///      - other_cond is `t1.col2 = t2.col2 and t1.col3 > t2.col3`
+///      - other_cond_from_in is `t1.col1 = t2.col1`
+///   2. anti semi join.
+///      - join keys are `t1.col2` and `t2.col2`
+///      - other_cond is `t1.col3 > t2.col3`
+///      - other_cond_from_in is `t1.col1 = t2.col1`
+///   3. null-aware anti semi join
+///      - join keys are `t1.col1` and `t2.col1`
+///      - other_cond is `t1.col2 = t2.col2 and t1.col3 > t2.col3`
+///      - null_aware_eq_cond is `t1.col1 = t2.col1`
+struct JoinNonEqualConditions
 {
+    String left_filter_column;
+    String right_filter_column;
```

Comment:
how about
```
    String probe_filter_column;
    String build_filter_column;
```

Reply:
I just keep the original name, I think it will make the code more understandable by using `left/right_filter_column`. For example
```
if (unlikely(!left_filter_column.empty() && !isLeftJoin(kind)))
            return "non left join with left conditions";
if (unlikely(!right_filter_column.empty() && !isRightJoin(kind)))
            return "non right join with right conditions";
```

Reply:
ok